File: app/src/main/java/com/maxandron/daggerissue325java/MyApplication.java
package com.maxandron.daggerissue325java;

import android.app.Application;
import android.content.Context;

/**
 * Created by maxandron on 25/02/16.
 */
public class MyApplication extends Application {
    public static MyApplication get(Context context) {
        return (MyApplication) context.getApplicationContext();
    }

    private MyComponent mComponent;

    @Override
    public void onCreate() {
        super.onCreate();

        mComponent = DaggerMyComponent.builder().myProvider(new MyProvider()).build();
    }

    public MyComponent getAppComponent() {
        return mComponent;
    }
}

File: app/src/main/java/com/maxandron/daggerissue325java/MyApplication.kt
package com.maxandron.daggerissue325java

import android.app.Application
import android.content.Context

class MyApplication : Application() {

    lateinit var appComponent: MyComponent

    override fun onCreate() {
        super.onCreate()

        appComponent = DaggerMyComponent.builder().myProvider(MyProvider()).build()
    }

    companion object {
        operator fun get(context: Context): MyApplication {
            return context.applicationContext as MyApplication
        }
    }
}
